### PR TITLE
Feature/lineup constraints

### PR DESCRIPTION
**Pull request by sharkiteuthis**

### CODE REVIEW

File: draftfast/optimizer.py [hunk 1]
```diff
@@ -56,6 +52,25 @@ def solve(self):
         solution = self.solver.Solve()
         return solution == self.solver.OPTIMAL
 
+    def _set_player_constraints(self):
```
Comment (BenBrostoff) [inline draftfast/optimizer.py]: Nice, like this refactor. One thing we should be mindful of is that for showdown, you may want the ability to ban a player as a captain but allow him to be picked for another spot. Not sure how to handle this in `Optimizer`, so let me know if you have ideas. It's also outside the scope of this PR so feel free to ignore this.

File: draftfast/test/test_ruleset.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+def test_multi_lock_multi_ban():
+    rs = RuleSet()
+    rs.add_group_rule(['Spencer Ware', 'Amari Cooper'], 2)
+    rs.add_group_rule(['Eli Manning', 'Russell Wilson', 'Doug Martin'], 0)
+
+    ntool.assert_equal(len(rs), 2)
+
+
+@ntool.raises(RuleException)
```
Comment (BenBrostoff) [inline draftfast/test/test_ruleset.py]: Didn't know about this decorator 👍

**Comment by sharkiteuthis** [comment]
this is ready except for tests, but the issue from #115 is blocking

File: draftfast/optimizer.py [hunk 3]
```diff
@@ -25,32 +29,53 @@ def __init__(
...
             self.player_to_idx_map[player.solver_id] = idx
+            self.name_to_idx_map[player.name] = idx
+
+            if self._is_locked(player):
+                player.lock = True
+            if self._is_banned(player):
+                player.ban = True
+
+            if player.lock and player.ban:
+                raise PlayerBanAndLockException(player.name)
```
Comment (BenBrostoff) [inline draftfast/optimizer.py]: 👍

File: draftfast/test/data/dk-nba-pickem-salaries.csv [hunk 4]
```diff
@@ -1,40 +1,40 @@
-Position,Name + ID,Name,ID,Roster Position,Game Info,TeamAbbrev,AvgPointsPerGame
```
Comment (BenBrostoff) [inline draftfast/test/data/dk-nba-pickem-salaries.csv]: What's the reason for the change in the diff here? I just took these directly from DraftKings.

File: draftfast/lineup_contraints.py [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+        return set(self.players) == set(rule.players)
+
+    def __hash__(self):
+        return hash(''.join(sorted(self.players)))
+
+    def __contains__(self, player):
+        return player in self.players
+
+
+class PlayerGroupConstraint(PlayerConstraint):
```
Comment (BenBrostoff) [inline draftfast/lineup_contraints.py]: This is cool how the abstract methods / inheritance works 👍

File: draftfast/optimizer.py [hunk 6]
```diff
@@ -61,6 +86,38 @@ def solve(self):
...
+    def _set_player_group_constraints(self):
+        for group_constraint in self.lineup_constraints:
+            if group_constraint.exact:
+                lb = ub = group_constraint.exact
+            else:
+                lb = group_constraint.lb
+                ub = group_constraint.ub
+
+            constraint = self.solver.Constraint(lb, ub)
+            for name in group_constraint.players:
```
Comment (BenBrostoff) [inline draftfast/optimizer.py]: The library consumer can still make this impossible by having some ridiculous group that has say three QBs and requests 2. If you do this right now, here's the error with only having constraints:

```
No solution found for command line query.
            Try adjusting your query by taking away constraints.

            Active constraints: None
            Player count: 784
```

So just to improve the messaging, I think you just need to modify `optimize` to maybe show the optimizer settings, the lineup constraints and the number of players in the pool (it also shouldn't say `command line query` which is my fault ). IMO the most frustrating thing I encounter when debugging this is why no solutions happen - this is usually user error and my hope is that this error message is helpful.

**Comment by sharkiteuthis** [comment]
I think there's a bug here with multi position players and groups, I'm going to try to reproduce and create a testcase

**Comment by BenBrostoff** [comment]
Yep, [it's this one](https://github.com/BenBrostoff/draftfast/blob/master/draftfast/test/test_nba_general.py#L46-L50). I think this started intermittently failing when I added the minimum 2 teams rule (which actually only really helps for Showdown, it's min 2 teams and max 4 players on one team for Classic). The roster tested here used to always be 3 guards, 4 forwards and 1 center. Sometimes this passes, but I've noticed occasionally it will be 4/3/1 or 3/3/2.

File: draftfast/test/data/dk-nba-pickem-salaries.csv (hunk 4 above)
Comment (sharkiteuthis) [reply draftfast/test/data/dk-nba-pickem-salaries.csv]: git strangeness. i must have accidentally deleted and re-added the file in different commits, i don't think the file actually changed

File: draftfast/optimizer.py (hunk 6 above)
Comment (sharkiteuthis) [reply draftfast/optimizer.py]: Here's what it looks like now:
```

No solution found.
Try adjusting your query by taking away constraints.

OPTIMIZER CONSTRAINTS:

Stacks: [('NE', 100)]
Min teams: 2
No offense against D: True

LINEUP CONSTRAINTS:

Using 2 of: Philip Rivers, Sam Bradford, Andrew Luck
Using 1 to 3 of: Saints, Alshon Jeffery, Lamar Miller
Banned: Sammy Watkins, Kellen Davis
Locked: Spencer Ware

PLAYER POOL SETTINGS:

None

PLAYER COUNT: 784
```

**Comment by sharkiteuthis** [comment]
@BenBrostoff once the builds pass this should be good to go

File: draftfast/optimize.py [hunk 7]
```diff
@@ -46,13 +51,28 @@ def run(rule_set: RuleSet,
...
-            No solution found for command line query.
-            Try adjusting your query by taking away constraints.
+No solution found.
+Try adjusting your query by taking away constraints.
 
-            Active constraints: {}
-            Player count: {}
-            '''
-            .format(optimizer_settings, len(players or []))
+OPTIMIZER CONSTRAINTS:
```
Comment (BenBrostoff) [inline draftfast/optimize.py]: Great stuff 👍
